src/objectExplorer/objectExplorerService.ts

```diff
@@ -130,44 +266,71 @@ export class ObjectExplorerService {
         return undefined;
     }
 
-    public async removeObjectExplorerNode(node: TreeNodeInfo): Promise<void> {
+    public async removeObjectExplorerNode(node: TreeNodeInfo, isDisconnect: boolean = false): Promise<void> {
         await this.closeSession(node);
-        const index = this._rootTreeNodeArray.indexOf(node, 0);
-        if (index > -1) {
-            this._rootTreeNodeArray.splice(index, 1);
+        if (!isDisconnect) {
+            const index = this._rootTreeNodeArray.indexOf(node, 0);
+            if (index > -1) {
+                this._rootTreeNodeArray.splice(index, 1);
+            }
         }
-        this._currentNode = undefined;
-        const nodeUri = node.nodePath + '_' + node.label;
+        const nodeUri = ObjectExplorerUtils.getNodeUri(node);
         this._connectionManager.disconnect(nodeUri);
-        await this._objectExplorerProvider.refresh(undefined);
+        this._treeNodeToChildrenMap.delete(node);
+        this._nodePathToNodeLabelMap.delete(node.nodePath);
+        this._sessionIdToConnectionCredentialsMap.delete(node.sessionId);
+        if (this._sessionIdToPromiseMap.has(node.sessionId)) {
+            this._sessionIdToPromiseMap.delete(node.sessionId);
+        }
+        this._currentNode = undefined;
+        if (isDisconnect) {
+            this._treeNodeToChildrenMap.set(node, [new ConnectTreeNode(node)]);
+            this.updateNode(node);
+            return this._objectExplorerProvider.refresh(undefined);
+        }
     }
 
-    public async refreshNode(node: TreeNodeInfo): Promise<boolean> {
+    public async refreshNode(node: TreeNodeInfo): Promise<void> {
         const refreshParams: RefreshParams = {
             sessionId: node.sessionId,
             nodePath: node.nodePath
         };
-        const response = await this._connectionManager.client.sendRequest(RefreshRequest.type, refreshParams);
-        return response;
+        await this._connectionManager.client.sendRequest(RefreshRequest.type, refreshParams);
+        return this._objectExplorerProvider.refresh(node);
     }
 
-    private async closeSession(node: TreeNodeInfo): Promise<void> {
-        const closeSessionParams: CloseSessionParams = {
-            sessionId: node.sessionId
-        };
-        const response = await this._connectionManager.client.sendRequest(CloseSessionRequest.type,
-            closeSessionParams);
-        if (response && response.success) {
-            this._sessionIdToConnectionCredentialsMap.delete(response.sessionId);
-            node.sessionId = undefined;
-            this._currentNode = node;
-            this._treeNodeToChildrenMap.set(this._currentNode, undefined);
-            this._currentNode.collapsibleState = TreeItemCollapsibleState.Collapsed;
+    public signInNodeServer(node: TreeNodeInfo): void {
```

Comment:
No there's no way to handle this. VSCode's API doesn't allow you to refresh specific nodes without refreshing the root node. Even the azure extension does error handling the same way. I worked on this specific issue all last week. This is also because how the objectExplorerService is written in tools service. It doesn't have a response event for OE requests, but has notifications. Since the VSCode API expects children every time the treeDataChanged event is fired, I ended up using Deferred promises to mimic correct Promise behavior. However, to force the nodes to show up, I have to refresh them. 

Not removing the node because removing from the OE also removes from saved connections.

We can go over this offline in detail if you want.